db/util_testing.go

```diff
@@ -0,0 +1,78 @@
+package db
+
+import (
+	"math"
+
+	"github.com/couchbase/gocb"
+	"github.com/couchbase/sync_gateway/base"
+)
+
+// Workaround SG #3570 by doing a polling loop until the star channel query returns 0 results.
+// Uses the star channel index as a proxy to indicate that _all_ indexes are empty (which might not be true)
+func WaitForIndexEmpty(bucket *base.CouchbaseBucketGoCB, bucketSpec base.BucketSpec) error {
+
+	retryWorker := func() (shouldRetry bool, err error, value interface{}) {
+
+		var results gocb.QueryResults
+
+		// Create the star channel query
+		starChannelQueryStatement := replaceSyncTokensQuery(QueryStarChannel.statement, bucketSpec.UseXattrs)
+		params := map[string]interface{}{}
+		params[QueryParamStartSeq] = 0
+		params[QueryParamEndSeq] = math.MaxInt64
+
+		// Execute the query
+		results, err = bucket.Query(starChannelQueryStatement, params, gocb.RequestPlus, true)
+
+		// If there was an error, then retry.  Assume it's an "index rollback" error which happens as
+		// the index processes the bucket flush operation
+		if err != nil {
+			base.Infof(base.KeyAll, "Error querying star channel: %v.  Assuming it's a temp error, will retry", err)
+			return true, err, nil
+		}
+
+		// Count how many results there are
+		count := CountQueryResults(results)
+
```

Comment:
Need to check for error on results.Close() before assuming we're successful.

Reply:
I added a check for error on `results.Close()`, but not I'm not exactly sure what to do of it.  For now, I'm just logging a warning, but let me know what you think the appropriate behavior should be here.

Reply:
I think results.Close() should be getting called when you finish iterating through the query results (and not on defer).  Row-level unmarshalling errors are only returned on Close().  I expect this should be treated the same as hitting an error when calling bucket.Query for this use case.

Reply:
Fixed in 46255eaef9f597b1ae3123dfd9fef199b56328d4